src/boot/memory.c

```diff
@@ -115,139 +111,126 @@ void move_segment_table_to_dmem(void) {
+}
+
 /**
- * Initialize the main memory pool. This pool is conceptually a pair of stacks
+ * Initialize the main memory pool. This pool is conceptually regions
  * that grow inward from the left and right. It therefore only supports
  * freeing the object that was most recently allocated from a side.
  */
-void main_pool_init(void *start, void *end) {
-    sPoolStart = (u8 *) ALIGN16((uintptr_t) start) + 16;
-    sPoolEnd = (u8 *) ALIGN16((uintptr_t) end - 15) - 16;
-    sPoolFreeSpace = sPoolEnd - sPoolStart;
-
-    sPoolListHeadL = (struct MainPoolBlock *) (sPoolStart - 16);
-    sPoolListHeadR = (struct MainPoolBlock *) sPoolEnd;
-    sPoolListHeadL->prev = NULL;
-    sPoolListHeadL->next = NULL;
-    sPoolListHeadR->prev = NULL;
-    sPoolListHeadR->next = NULL;
-#ifdef PUPPYPRINT_DEBUG
-    mempool = sPoolFreeSpace;
+void main_pool_init() {
+    main_pool_region_init(&gMainPool.regions[0], __mainPoolStart, (void*) RAM_END);
+}
+
+// This function behaves similar to main_pool_region_alloc_from_start but it is used for freeable memory allocated from the end.
+#define main_pool_region_alloc_from_end_freeable(region, size, alignment, try) _main_pool_region_alloc_from_end_freeable(region, size, alignment, try, __LINE__, __FILE__)
+static ALWAYS_INLINE void* _main_pool_region_alloc_from_end_freeable(int region_idx, u32 size, s32 alignment, int try, int line, const char* file) {
+    struct MainPoolRegion* region = &gMainPool.regions[region_idx];
+    u8* region_end = region->end;
+    u8* new_end;
+    if (alignment > 0)
+        new_end = (u8*) DOWN(region_end - size, alignment) - sizeof(struct MainPoolFreeableHeader);
+    else
+        new_end = region_end - size - sizeof(struct MainPoolFreeableHeader);
+
+    if (try) {
+        if (new_end < region->start)
+            return NULL;
+    } else {
+#ifdef DEBUG_MEMORY_ALLOC
+        if (new_end < region->start)
+            main_pool_alloc_failed(region_idx, size, TRUE, line, file);
+#else
+        (void) line; (void) file;
 #endif
+    }
+
+    struct MainPoolFreeableHeader* header = (struct MainPoolFreeableHeader*) new_end;
+    header->magic = MAIN_POOL_FREEABLE_HEADER_MAGIC_RIGHT;
+    header->id = region_idx;
+    header->ptr = region_end;
+
+    region->end = new_end;
+
+    return header->data;
 }
 
-/**
- * Allocate a block of memory from the pool of given size, and from the
- * specified side of the pool (MEMORY_POOL_LEFT or MEMORY_POOL_RIGHT).
- * If there is not enough space, return NULL.
- */
-void *main_pool_alloc(u32 size, u32 side) {
-    struct MainPoolBlock *newListHead;
-    void *addr = NULL;
+void main_pool_alloc_failed(int region_idx, u32 size, int freeable, int line, const char* file)
+{
+    struct MainPoolRegion* region = &gMainPool.regions[region_idx];
+    char errmsg[70];
+    sprintf(errmsg, "%salloc failed for %d bytes in region %d\n%x-%x", freeable ? "freeable " : "", size, region_idx, region->start, region->end);
+    __n64Assert((char*) file, line, errmsg);
```

Comment:
nit: ~~use the `error` macro to simplify this (and to make replacing these with `assert_args` or similar easier to find once 2.4 gets merged into 3.0)~~

Nevermind, I see what's going on here. Add a TODO then for replacing `errmsg` with `gAssertionStr` once 2.4 is merged.

Reply:
Added todo, waiting for 2.4 merge...